File: DKF-Python/InitializeGradsAndMoments.py
import numpy as np
import copy
def InitializeGradsAndMoments(NetWeights, NetParameters):
    """
    Initializes the gradients for the net's parameters to zero.
    """
    Layers = NetParameters['Layers']
    SharedWeights = NetParameters['SharedWeights']
    HiddenDynamicsNumber = NetParameters['HiddenDynamicsNumber']

    # Setup gradients
    if SharedWeights == 'No':
        Grads = [None] * (Layers + 1)

        for Layer in range(Layers):
            Grads[Layer] = np.zeros_like(NetWeights[Layer]).astype('float64')
        
        Grads[Layers] = [None] * (HiddenDynamicsNumber + 1)
        for Dyn in range(HiddenDynamicsNumber):
            Grads[Layers][Dyn] = np.zeros_like(NetWeights[Layers][Dyn]).astype('float64')
        Moment1 = copy.deepcopy(Grads)
        Moment2 = copy.deepcopy(Grads)

    if SharedWeights == 'Yes':
        Grads = [None] * (2)

        Grads[0] = np.zeros_like(NetWeights[0]).astype('float64')

        Grads[1] = [None] * (HiddenDynamicsNumber + 1)
        for Dyn in range(HiddenDynamicsNumber):
            Grads[1][Dyn] = np.zeros_like(NetWeights[1][Dyn]).astype('float64')
        Moment1 = copy.deepcopy(Grads)
        Moment2 = copy.deepcopy(Grads)

    return Grads, Moment1, Moment2

File: DKF-MATLAB/InitializeGradsAndMoments.m
function [Grads,Moment1,Moment2] = InitializeGradsAndMoments(NetWeights,NetParameters)
%INITIALIZEGRADSANDMOMENTS: Initializes the gradients for the net's parameters to zero.

%Variables
Layers = NetParameters.Layers;
SharedWeights = NetParameters.SharedWeights;
HiddenDynamicsNumber = NetParameters.HiddenDynamicsNumber;

%Setup gradients
if strcmp(SharedWeights,'Yes')
    Grads = cell(1,2);

    Grads{1} = zeros(size(NetWeights{1}));
    for Dyn = 1:HiddenDynamicsNumber
        Grads{2}{Dyn} = zeros(size(NetWeights{2}{Dyn}));
    end
end

if strcmp(SharedWeights,'No')
    Grads = cell(1,Layers+1);

    for Layer = 1:Layers
        Grads{Layer} = zeros(size(NetWeights{Layer}));
    end
    for Dyn = 1:HiddenDynamicsNumber
        Grads{Layers+1}{Dyn} = zeros(size(NetWeights{Layers+1}{Dyn}));
    end
end

if nargout > 1
    Moment1 = Grads;
    Moment2 = Grads;
end
end

